### PR TITLE
infra: Bump Go to version 1.20.12

### PR DESCRIPTION
**Pull request by brianmcarey**
**What this PR does / why we need it**:

There are a number of security fixes that are included in Go 1.20.12 [1] - KubeVirt should take advantage of these security improvements.

[1] https://go.dev/doc/devel/release#go1.20 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- ~[ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required~
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- ~[ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)~
- ~[ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required~
- ~[ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test~
- ~[ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.~
- ~[ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered~

**Release note**:

```release-note
Built with golang 1.20.12
```

### CODE REVIEW

**Comment by xpivarc** [comment]
There are some concerning tests that failed but otherwise, I think this is going in the right direction. @brianmcarey Thanks Brian I know this one was not easy :) I will have a look at the failures, hopefully this week.

**Comment by xpivarc** [comment]
/retest
@EdDev Could you have a look at the failing test?

**Comment by EdDev** [comment]
> /retest @EdDev Could you have a look at the failing test?

I'm seeing a problem with running the sidecar: https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/10879/pull-kubevirt-e2e-k8s-1.28-sig-network/1736770984142180352/artifacts/k8s-reporter/pods/1_kubevirt-test-default1_virt-launcher-testvmi-2mhmp-qdhtd-hook-sidecar-0.log

**Comment by EdDev** [comment]
> Built with golang 1.20.12

Any reason we did not go directly to 1.21 ?

**Comment by brianmcarey** [comment]
> > Built with golang 1.20.12
> 
> Any reason we did not go directly to 1.21 ?

Kubernetes have just started building against 1.20.12[1] and I think it makes sense to stay inline with Kubernetes. 

We also need to move off of 1.19 and I hit a couple of issues with 1.21 so I think for now its best to upgrade to 1.20.12. 

[1] https://github.com/kubernetes/kubernetes/pull/122216

File: BUILD.bazel [hunk 1]
```diff
@@ -54,10 +54,10 @@ nogo(
     # that are always correct).
     # You can see the what `go vet` does by running `go doc cmd/vet`.
     deps = [
-        "//vendor/github.com/nunnatsa/ginkgolinter:go_default_library",
+        #"//vendor/github.com/nunnatsa/ginkgolinter:go_default_library",
```
Comment (dhiller) [inline BUILD.bazel:57]: We should keep track of disabling the nogo plugins, and reenable them at a later point in time.

File: WORKSPACE [hunk 2]
```diff
@@ -44,23 +44,28 @@ rules_proto_toolchains()
...
     urls = [
-        "https://mirror.bazel.build/github.com/bazelbuild/rules_go/releases/download/v0.35.0/rules_go-v0.35.0.zip",
-        "https://github.com/bazelbuild/rules_go/releases/download/v0.35.0/rules_go-v0.35.0.zip",
-        "https://storage.googleapis.com/builddeps/099a9fb96a376ccbbb7d291ed4ecbdfd42f6bc822ab77ae6f1b5cb9e914e94fa",
+        "https://mirror.bazel.build/github.com/bazelbuild/rules_go/releases/download/v0.42.0/rules_go-v0.42.0.zip",
+        "https://github.com/bazelbuild/rules_go/releases/download/v0.42.0/rules_go-v0.42.0.zip",
+    ],
+)
+
+http_archive(
```
Comment (dhiller) [inline WORKSPACE:54]: We should add a `FIXME` note here for why this is necessary, since we might want to remove it at a later point in time.

Comment (brianmcarey) [reply WORKSPACE:55]: I've added notes in the latest push.

**Review by dhiller — APPROVED** [review]
Thank you @brianmcarey, great work!

/lgtm

**Comment by EdDev** [comment]
> Kubernetes have just started building against 1.20.12[1] and I think it makes sense to stay inline here.
> 
> We also need to move off of 1.19 and I hit a couple of issues with 1.21 so I think for now its best to upgrade to 1.20.12.
> 
> [1] [kubernetes/kubernetes#122216](https://github.com/kubernetes/kubernetes/pull/122216)

I think our main branch is targeting to integrate with k8s 1.29, right?
If so, then k8s uses Go 1.21 already [1].

Perhaps it is worth to continue with this great effort now that everything is “fresh”.

[1] https://github.com/kubernetes/kubernetes/pull/122201

**Comment by EdDev** [comment]
I see some (or all?) of the nogo linters have been disabled.

This is a problem.

- Could you clarify which linters we lost?
- Can we use the “regular” linter job as the alternative and mark it “required”? The lost linters will have to be added there.

**Comment by brianmcarey** [comment]
> > Kubernetes have just started building against 1.20.12[1] and I think it makes sense to stay inline here.
> > We also need to move off of 1.19 and I hit a couple of issues with 1.21 so I think for now its best to upgrade to 1.20.12.
> > [1] [kubernetes/kubernetes#122216](https://github.com/kubernetes/kubernetes/pull/122216)
> 
> I think our main branch is targeting to integrate with k8s 1.29, right? If so, then k8s uses Go 1.21 already [1].
> 
> Perhaps it is worth to continue with this great effort now that everything is “fresh”.

The update to go 1.21[2] is much more straight forward thanks to the changes in this PR so the option is there if we just want to go to 1.21.5 - I missed that 1.29 was being built with 1.21. 

[2] https://github.com/kubevirt/kubevirt/pull/10914
>

**Pull request merged**